sdk/core/core/src/az_http_policy.c

```diff
@@ -77,6 +79,34 @@ AZ_NODISCARD az_result az_http_pipeline_policy_credential(
   return az_http_pipeline_nextpolicy(p_policies, p_request, p_response);
 }
 
+AZ_NODISCARD az_result az_http_pipeline_policy_header_validation(
+    _az_http_policy* p_policies,
+    void* p_data,
+    _az_http_request* p_request,
+    az_http_response* p_response)
+{
+  (void)p_data;
+  az_result call_next_policy_result
+      = az_http_pipeline_nextpolicy(p_policies, p_request, p_response);
+
+  // Validate headers on response
+  // 1. parse status line to have access to headers
+  az_http_response_status_line status = { 0 };
+  AZ_RETURN_IF_FAILED(az_http_response_get_status_line(p_response, &status));
+
+  // loop headers and validate
+  for (az_pair header;
+       az_http_response_get_next_header(p_response, &header) != AZ_ERROR_ITEM_NOT_FOUND;)
+  {
+    if (!az_http_is_valid_header_name(header.key))
```

Comment:
This causes double parsing, which is a bit unfortunate. Can we design it differently to avoid the validation overhead?

Reply:
Well, it depends. 
Let me tag @JeffreyRichter  for this.

So, If we want to fail request if the response comes with an invalid header, and without making a double parsing, one option is to build a new API for httpResponse that transport-layer would use to write the response.
Right now we only provide an az_span to transport-layer and we delegate the responsibility of writing a valid HttpResponse to it. So, if there is any bug on any transport layer to write a valid RFC response (status lines, headers or even body).
Now, consider adding next API and hiding the az_span buffer from transport-layer
```c
az_http_response_write_status_line(httpResponse *r, az_span status_line);
az_http_response_append_header(httpResponse *r, az_span header);
az_http_response_append_chuck_body(httpResponse *r, az_span bodyChunk);
```
If we decide to add this, we would be taking control of the validation and correctness of any transport-layer. Then we can fail the request as soon as trasnport-layer gives us an invalid header.
Downside is that we need to force any transport-layer to always write status-line first, then headers and finally a body (to maintain the valid structure)

Reply:
I don't think we need to invalidate the entire response.